functions/go/set-namespace/transformer/namespace.go

```diff
@@ -18,206 +18,311 @@ import (
+			nsPtr := &name
+			visitor(o.GetOriginId().Name, nsPtr, o.ShortString())
+			o.SetName(*nsPtr)
+		case o.IsGVK("apiextensions.k8s.io", "v1", "CustomResourceDefinition"):
+			namespace := o.NestedStringOrDie("spec", "conversion", "webhook", "clientConfig", "service", "namespace")
+			nsPtr := &namespace
+			visitor("", nsPtr)
+			o.SetNestedStringOrDie(*nsPtr, "spec", "conversion", "webhook", "clientConfig", "service", "namespace")
+		case o.IsGVK("apiregistration.k8s.io", "v1", "APIService"):
+			namespace := o.NestedStringOrDie("spec", "service", "namespace")
+			nsPtr := &namespace
+			visitor("", nsPtr)
+			o.SetNestedStringOrDie(*nsPtr, "spec", "service", "namespace")
+		case o.GetKind() == "ClusterRoleBinding" || o.GetKind() == "RoleBinding":
+			subjects := o.GetSlice("subjects")
+			for _, s := range subjects {
+				namespace := s.NestedStringOrDie("namespace")
+				nsPtr := &namespace
+				visitor("", nsPtr)
+				s.SetNestedStringOrDie(*nsPtr, "namespace")
 			}
+		default:
+			// skip the cluster scoped resource
 		}
-	})
+	}
 }
 
-type NamespaceTransformer struct {
-	NewNamespace string
-	DependsOnMap map[string]bool
-	Errors       []string
+// VisitNamespaceResource applies "visitor" to namespace-scoped resource.
+// We made a hypothesis here that if a unknown scoped resource has a non-empty metadata.namespace, the resource will be
+// treated as namespace scoped.
+func VisitNamespaceResource(objects fn.KubeObjects, visitor func(origin string, currentPtr *string, idStr ...string)) {
+	namespaceScoped := objects.Where(func(o *fn.KubeObject) bool { return o.IsNamespaceScoped() })
+	for _, o := range namespaceScoped {
+		namespace := o.GetNamespace()
+		nsPtr := &namespace
+		visitor(o.GetOriginId().Namespace, nsPtr, o.ShortString())
+		o.SetNamespace(*nsPtr)
+	}
 }
 
-func NewNamespace(obj *fn.KubeObject, namespacePtr *string) *Namespace {
-	annotationSetter := func(newAnnotation string) {
-		obj.SetAnnotation(dependsOnAnnotation, newAnnotation)
-	}
-	annotationGetter := func() string {
-		group := obj.GetAPIVersion()
-		if i := strings.Index(obj.GetAPIVersion(), "/"); i > -1 {
-			group = group[:i]
+// MapGKNNBeforeChange stores each namespace-scoped resource's Group, Kind, Namespace and Name.
+// This map will be used later to align the depends-on annotation.
+func MapGKNNBeforeChange(objects fn.KubeObjects) map[string]struct{} {
+	dependsOnMap := map[string]struct{}{}
+	for _, o := range objects {
+		id := o.GetId()
+		if id.Namespace == fn.UnknownNamespace {
+			continue
 		}
-		return dependsOnKeyPattern(group, obj.GetKind(), obj.GetName())
-	}
-	return &Namespace{
-		Ptr:                 namespacePtr, //  obj.GetStringOrDie(path...),
-		IsNamespace:         obj.IsGVK("v1", "Namespace"),
-		DependsOnAnnotation: obj.GetAnnotations()[dependsOnAnnotation],
-		annotationGetter:    annotationGetter,
-		annotationSetter:    annotationSetter,
+		key := fmt.Sprintf("%v/namespaces/%v/%v/%v", id.Group, id.Namespace, id.Kind, id.Name)
```

Comment:
It might be useful to have a helper to format a key string form an `id` object.
I'm curious why use choose this format as the key.

Reply:
..same here..... it's a little bit error-prone .